funcs_utils.php

```diff
@@ -0,0 +1,289 @@
+    $modules = [];
+    foreach ($json as $module) {
+        $ghrepo = $module['github'];
+        $modules[] = [
+            'ghrepo' => $ghrepo,
+            'account' => explode('/', $ghrepo)[0],
+            'repo' => explode('/', $ghrepo)[1],
+            'cloneUrl' => "git@github.com:$ghrepo.git",
+            'branch' => max($module['branches'] ?: [-1])
+        ];
+    }
+    return $modules;
+}
+
+/**
+ * Returns a list of all scripts files to run against a particular cms major version
+ */
+function script_files($cmsMajor)
+{
+    if (!ctype_digit($cmsMajor)) {
+        $cmsMajor = "-$cmsMajor";
+    }
+    $scriptFiles = [];
+    $dir = "scripts/cms$cmsMajor";
+    if (!file_exists($dir)) {
+        warning("$dir does not exist, no CMS $cmsMajor specific scripts will be run");
+        return $scriptFiles;
+    }
+    if ($handle = opendir($dir)) {
+        while (false !== ($scriptFile = readdir($handle))) {
+            if ('.' === $scriptFile || '..' === $scriptFile) {
+                continue;
+            }
+            $scriptFiles[] = "$dir/$scriptFile";
+        }
+        closedir($handle);
+    }
+    return $scriptFiles;
+}
+
+/**
+ * Runs a shell command and returns the output
+ */
+function cmd($cmd, $cwd)
+{
+    info("Running command: $cmd in $cwd");
+    // using Process::fromShellCommandline() instead of new Process() so that pipes work
+    $process = Process::fromShellCommandline($cmd, $cwd);
+    $process->run();
+    if (!$process->isSuccessful()) {
+        warning("Error running command: $cmd in $cwd");
+        error("Output was: " . $process->getErrorOutput());
+    }
+    return trim($process->getOutput());
+}
+
+/**
+ * Returns a object used to output to the console
+ */
+function io(): SymfonyStyle
+{
+    global $IN, $OUT;
+    return new SymfonyStyle($IN ?: new ArgvInput(), $OUT ?: new NullOutput);
+}
+
+/**
+ * Removes a directory
+ */
+function remove_dir($dirname)
+{
+    if (!file_exists(($dirname))) {
+        return;
+    }
+    info("Removing $dirname");
+    shell_exec("rm -rf $dirname");
+}
+
+/**
+ * Validates the users system is ready to run the script
+ */
+function validate_system()
+{
+    if (!github_token()) {
+        error('Could not get github token - set MS_GITHUB_TOKEN environment variable');
+    }
+}
+
+/**
+ * Reads MS_GITHUB_TOKEN environment variable
+ */
+function github_token()
+{
+    return getenv('MS_GITHUB_TOKEN') ?: '';
+}
+
+/**
+ * Makes a request to the github API
+ */
+function github_api($url, $data = [])
+{
+    $token = github_token();
+    $jsonStr = empty($data) ? '' : json_encode($data, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE);
+    $ch = curl_init($url);
+    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
+    curl_setopt($ch, CURLOPT_POST, !empty($data));
+    curl_setopt($ch, CURLOPT_HTTPHEADER, [
+        'User-Agent: silverstripe-module-standardiser',
+        'Accept: application/vnd.github+json',
+        "Authorization: Bearer $token",
```

Comment:
It might trigger a secondary rate limit even with the token, this program will output the modules that were processed which you can then copy pate into an --exclude command line flag that makes it easier to pickup where you left off